src/ripple/app/ledger/impl/LedgerMaster.cpp

```diff
@@ -1514,11 +1522,125 @@ LedgerMaster::shouldAcquire (
     return ret;
 }
 
+void
+LedgerMaster::fetchForHistory(
+    std::uint32_t missing,
+    bool& progress,
+    InboundLedger::Reason reason)
+{
+    ScopedUnlockType sl(m_mutex);
+    if (auto hash = getLedgerHashForHistory(missing, reason))
+    {
+        assert(hash->isNonZero());
+        auto ledger = getLedgerByHash(*hash);
+        if (! ledger)
+        {
+            if (!app_.getInboundLedgers().isFailure(*hash))
+            {
+                ledger = app_.getInboundLedgers().acquire(
+                    *hash, missing, reason);
+                if (!ledger && (missing > 32600) && missing != fetch_seq_)
+                {
+                    JLOG(m_journal.trace())
+                        << "fetchForHistory want fetch pack " << missing;
+                    fetch_seq_ = missing;
```

Comment:
`fetch_seq` simply tracks the ledger sequence for the last fetch pack requested. Its purpose is to prevent requesting the same fetch more than once. Generally, more than one ledger is fetched at the same time. For instance, if the ledger isn't present locally and there isn't a fetch pack available, we request the prior ten ledgers at once.

Reply:
So if someone calls 
```
fetchForHistory(2);
fetchForHistory(1);
fetchForHistory(2);
fetchForHistory(1);
```
Does that means multiple fetches will happen for the same ledger?

Reply:
I am glad you asked as it made me realize `fetchForHistory` should be private! The code in the function was originally in `doAdvance` and I broke it out as it was too tall. `DoAdvance` is called from the jobQueue and it can and will call `fetchForHistory` with a prior sequence. That is by design and it won't hurt anything as inboundledgers::acquire will only have one fetch operation per ledger.